File: app/src/main/java/org/breezyweather/background/interfaces/CMWeatherProviderService.java
package org.breezyweather.background.interfaces;

import android.text.TextUtils;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import cyanogenmod.providers.WeatherContract;
import cyanogenmod.weather.RequestInfo;
import cyanogenmod.weather.WeatherInfo;
import cyanogenmod.weatherservice.ServiceRequest;
import cyanogenmod.weatherservice.ServiceRequestResult;
import cyanogenmod.weatherservice.WeatherProviderService;
import dagger.hilt.android.AndroidEntryPoint;
import org.breezyweather.common.basic.models.Location;
import org.breezyweather.common.basic.models.options.provider.WeatherSource;
import org.breezyweather.common.basic.models.weather.Weather;
import org.breezyweather.common.basic.models.weather.WeatherCode;
import org.breezyweather.location.LocationHelper;
import org.breezyweather.main.utils.RequestErrorType;
import org.breezyweather.settings.SettingsManager;
import org.breezyweather.weather.WeatherHelper;

/**
 * CM weather provider service.
 * */

@AndroidEntryPoint
public class CMWeatherProviderService extends WeatherProviderService
        implements WeatherHelper.OnRequestWeatherListener {

    @Nullable
    private ServiceRequest mRequest;

    @Inject LocationHelper mLocationHelper;
    @Inject WeatherHelper mWeatherHelper;

    private final LocationHelper.OnRequestLocationListener locationListener
            = new LocationHelper.OnRequestLocationListener() {

        @Override
        public void requestLocationSuccess(Location requestLocation) {
            if (mRequest != null) {
                mWeatherHelper.requestWeather(
                        CMWeatherProviderService.this,
                        requestLocation,
                        CMWeatherProviderService.this);
            }
        }

        @Override
        public void requestLocationFailed(Location requestLocation, RequestErrorType requestErrorType) {
            if (mRequest != null) {
                mRequest.fail();
            }
        }
    };

    private final WeatherHelper.OnRequestLocationListener weatherLocationListener
            = new WeatherHelper.OnRequestLocationListener() {
        @Override
        public void requestLocationSuccess(String query, List<Location> locationList) {
            if (mRequest != null) {
                if (locationList != null && locationList.size() > 0) {
                    mWeatherHelper.requestWeather(
                            CMWeatherProviderService.this,
                            locationList.get(0),
                            CMWeatherProviderService.this);
                } else {
                    requestLocationFailed(query, RequestErrorType.WEATHER_REQ_FAILED);
                }
            }
        }

        @Override
        public void requestLocationFailed(String query, RequestErrorType requestErrorType) {
            if (mRequest != null) {
                mRequest.fail();
            }
        }
    };

    @Override
    public void onCreate() {
        super.onCreate();
        mRequest = null;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        cancelRequest();
    }

    @Override
    protected void onRequestSubmitted(ServiceRequest serviceRequest) {
        cancelRequest();
        mRequest = serviceRequest;

        RequestInfo info = serviceRequest.getRequestInfo();
        switch (info.getRequestType()) {
            case RequestInfo.TYPE_WEATHER_BY_WEATHER_LOCATION_REQ:
                requestWeather(info.getWeatherLocation().getCity());
                break;

            case RequestInfo.TYPE_WEATHER_BY_GEO_LOCATION_REQ:
                requestLocation();
                break;

            case RequestInfo.TYPE_LOOKUP_CITY_NAME_REQ:
                requestWeather(info.getCityName());
                break;

            default:
                serviceRequest.fail();
                break;
        }
    }

    @Override
    protected void onRequestCancelled(ServiceRequest serviceRequest) {
        cancelRequest();
    }

    // control.

    private void requestLocation() {
        mLocationHelper.requestLocation(
                this,
                Location.buildLocal(this),
                true,
                locationListener
        );
    }

    private void requestWeather(String cityName) {
        if (!TextUtils.isEmpty(cityName)) {
            WeatherSource weatherSource = SettingsManager.getInstance(this).getWeatherSource();
            mWeatherHelper.requestSearchLocations(this, cityName, weatherSource, weatherLocationListener);
        } else if (mRequest != null) {
            mRequest.fail();
        }
    }

    private void cancelRequest() {
        mRequest = null;
        mLocationHelper.cancel();
        mWeatherHelper.cancel();
    }

    // interface.

    // on request weather listener.

    @Override
    public void requestWeatherSuccess(@NonNull Location requestLocation) {
        try {
            Weather weather = requestLocation.getWeather();
            if (mRequest != null && weather != null) {
                List<WeatherInfo.DayForecast> forecastList = new ArrayList<>();
                for (int i = 0; i < weather.getDailyForecast().size(); i++) {
                    forecastList.add(
                            new WeatherInfo.DayForecast.Builder(
                                    WeatherConditionConvertHelper.getConditionCode(
                                            weather.getDailyForecast().get(i).getDay().getWeatherCode(),
                                            true
                                    ))
                                    .setHigh(weather.getDailyForecast().get(i).getDay().getTemperature().getTemperature())
                                    .setHigh(weather.getDailyForecast().get(i).getNight().getTemperature().getTemperature())
                                    .build()
                    );
                }
                WeatherInfo.Builder builder = new WeatherInfo.Builder(
                        requestLocation.getCityName(getApplicationContext()),
                        weather.getCurrent().getTemperature().getTemperature(),
                        WeatherContract.WeatherColumns.TempUnit.CELSIUS
                ).setWeatherCondition(
                                WeatherConditionConvertHelper.getConditionCode(
                                        weather.getCurrent().getWeatherCode(),
                                        requestLocation.isDaylight()
                                )
                ).setTodaysHigh(
                        weather.getDailyForecast().get(0).getDay().getTemperature().getTemperature()
                ).setTodaysLow(
                        weather.getDailyForecast().get(0).getNight().getTemperature().getTemperature()
                ).setTimestamp(
                        weather.getBase().getUpdateDate().getTime()
                );
                if (weather.getCurrent().getRelativeHumidity() != null) {
                    builder.setHumidity(weather.getCurrent().getRelativeHumidity());
                }
                if (weather.getCurrent().getWind().getSpeed() != null) {
                    builder.setWind(
                            weather.getCurrent().getWind().getSpeed(),
                            weather.getCurrent().getWind().getDegree().getDegree(),
                            WeatherContract.WeatherColumns.WindSpeedUnit.KPH
                    ).setForecast(forecastList);
                }

                mRequest.complete(new ServiceRequestResult.Builder(builder.build()).build());
            }
        } catch (Exception ignore) {
            requestWeatherFailed(requestLocation, RequestErrorType.WEATHER_REQ_FAILED);
        }
    }

    @Override
    public void requestWeatherFailed(@NonNull Location requestLocation, @NonNull RequestErrorType requestErrorType) {
        if (mRequest != null) {
            mRequest.fail();
        }
    }
}

class WeatherConditionConvertHelper {
    static int getConditionCode(WeatherCode code, boolean dayTime) {
        switch (code) {
            case CLEAR:
                if (dayTime) {
                    return WeatherContract.WeatherColumns.WeatherCode.SUNNY;
                } else {
                    return WeatherContract.WeatherColumns.WeatherCode.CLEAR_NIGHT;
                }

            case PARTLY_CLOUDY:
                if (dayTime) {
                    return WeatherContract.WeatherColumns.WeatherCode.PARTLY_CLOUDY_DAY;
                } else {
                    return WeatherContract.WeatherColumns.WeatherCode.PARTLY_CLOUDY_NIGHT;
                }

            case CLOUDY:
                return WeatherContract.WeatherColumns.WeatherCode.CLOUDY;

            case RAIN:
                return WeatherContract.WeatherColumns.WeatherCode.SHOWERS;

            case SNOW:
                return WeatherContract.WeatherColumns.WeatherCode.SNOW;

            case WIND:
                return WeatherContract.WeatherColumns.WeatherCode.WINDY;

            case FOG:
                return WeatherContract.WeatherColumns.WeatherCode.FOGGY;

            case HAZE:
                return WeatherContract.WeatherColumns.WeatherCode.HAZE;

            case SLEET:
                return WeatherContract.WeatherColumns.WeatherCode.SLEET;

            case HAIL:
                return WeatherContract.WeatherColumns.WeatherCode.HAIL;

            case THUNDER:
                return WeatherContract.WeatherColumns.WeatherCode.THUNDERSTORMS;

            case THUNDERSTORM:
                return WeatherContract.WeatherColumns.WeatherCode.THUNDERSHOWER;
        }
        return WeatherContract.WeatherColumns.WeatherCode.NOT_AVAILABLE;
    }
}

File: app/src/main/java/org/breezyweather/background/interfaces/CMWeatherProviderService.kt
package org.breezyweather.background.interfaces

import cyanogenmod.providers.WeatherContract
import cyanogenmod.weather.RequestInfo
import cyanogenmod.weather.WeatherInfo
import cyanogenmod.weather.WeatherInfo.DayForecast
import cyanogenmod.weatherservice.ServiceRequest
import cyanogenmod.weatherservice.ServiceRequestResult
import cyanogenmod.weatherservice.WeatherProviderService
import dagger.hilt.android.AndroidEntryPoint
import org.breezyweather.common.basic.models.Location
import org.breezyweather.common.basic.models.weather.WeatherCode
import org.breezyweather.location.LocationHelper
import org.breezyweather.main.utils.RequestErrorType
import org.breezyweather.settings.SettingsManager
import org.breezyweather.weather.WeatherHelper
import org.breezyweather.weather.WeatherHelper.OnRequestWeatherListener
import javax.inject.Inject

/**
 * CM weather provider service.
 */
@AndroidEntryPoint
class CMWeatherProviderService : WeatherProviderService(), OnRequestWeatherListener {
    private var mRequest: ServiceRequest? = null
    var mLocationHelper: LocationHelper? = null
        @Inject set
    var mWeatherHelper: WeatherHelper? = null
        @Inject set

    private val locationListener: LocationHelper.OnRequestLocationListener =
        object : LocationHelper.OnRequestLocationListener {
            override fun requestLocationSuccess(requestLocation: Location) {
                if (mRequest != null) {
                    mWeatherHelper!!.requestWeather(
                        this@CMWeatherProviderService,
                        requestLocation,
                        this@CMWeatherProviderService
                    )
                }
            }

            override fun requestLocationFailed(requestLocation: Location, requestErrorType: RequestErrorType) {
                mRequest?.fail()
            }
        }
    private val weatherLocationListener: WeatherHelper.OnRequestLocationListener =
        object : WeatherHelper.OnRequestLocationListener {
            override fun requestLocationSuccess(query: String, locationList: List<Location>) {
                if (mRequest != null) {
                    if (locationList.isNotEmpty()) {
                        mWeatherHelper!!.requestWeather(
                            this@CMWeatherProviderService,
                            locationList[0],
                            this@CMWeatherProviderService
                        )
                    } else {
                        requestLocationFailed(query, RequestErrorType.WEATHER_REQ_FAILED)
                    }
                }
            }

            override fun requestLocationFailed(query: String, requestErrorType: RequestErrorType) {
                mRequest?.fail()
            }
        }

    override fun onCreate() {
        super.onCreate()
        mRequest = null
    }

    override fun onDestroy() {
        super.onDestroy()
        cancelRequest()
    }

    override fun onRequestSubmitted(serviceRequest: ServiceRequest) {
        cancelRequest()
        mRequest = serviceRequest
        val info = serviceRequest.requestInfo
        when (info.requestType) {
            RequestInfo.TYPE_WEATHER_BY_WEATHER_LOCATION_REQ -> requestWeather(info.weatherLocation.city)
            RequestInfo.TYPE_WEATHER_BY_GEO_LOCATION_REQ -> requestLocation()
            RequestInfo.TYPE_LOOKUP_CITY_NAME_REQ -> requestWeather(info.cityName)
            else -> serviceRequest.fail()
        }
    }

    override fun onRequestCancelled(serviceRequest: ServiceRequest) {
        cancelRequest()
    }

    // control.
    private fun requestLocation() {
        mLocationHelper!!.requestLocation(
            this,
            Location.buildLocal(this),
            true,
            locationListener
        )
    }

    private fun requestWeather(cityName: String) {
        if (cityName.isNotEmpty()) {
            val weatherSource = SettingsManager.getInstance(this).weatherSource
            mWeatherHelper!!.requestSearchLocations(this, cityName, weatherSource, weatherLocationListener)
        } else if (mRequest != null) {
            mRequest!!.fail()
        }
    }

    private fun cancelRequest() {
        mRequest = null
        mLocationHelper!!.cancel()
        mWeatherHelper!!.cancel()
    }

    // interface.
    // on request weather listener.
    override fun requestWeatherSuccess(requestLocation: Location) {
        try {
            val weather = requestLocation.weather
            if (mRequest != null && weather?.current?.temperature?.temperature != null) {
                val forecastList: MutableList<DayForecast> = ArrayList()
                weather.dailyForecast.forEach { daily ->
                    forecastList.add(
                        DayForecast.Builder(
                            WeatherConditionConvertHelper.getConditionCode(daily.day?.weatherCode, true)
                        ).apply {
                            daily.day?.temperature?.temperature?.toDouble()?.let { setHigh(it) }
                            daily.night?.temperature?.temperature?.toDouble()?.let { setLow(it) }
                        }.build()
                    )
                }
                val builder = WeatherInfo.Builder(
                    requestLocation.getCityName(applicationContext),
                    weather.current.temperature.temperature.toDouble(),
                    WeatherContract.WeatherColumns.TempUnit.CELSIUS
                ).apply {
                    setWeatherCondition(
                        WeatherConditionConvertHelper.getConditionCode(
                            weather.current.weatherCode,
                            requestLocation.isDaylight
                        )
                    )
                    weather.dailyForecast.getOrNull(0)?.day?.temperature?.temperature?.toDouble()?.let { setTodaysHigh(it) }
                    weather.dailyForecast.getOrNull(0)?.night?.temperature?.temperature?.toDouble()?.let { setTodaysLow(it) }
                    setTimestamp(weather.base.updateDate.time)
                    weather.current.relativeHumidity?.toDouble()?.let { setHumidity(it) }
                    if (weather.current.wind?.speed != null && weather.current.wind.degree?.degree != null) {
                        setWind(
                            weather.current.wind.speed.toDouble(),
                            weather.current.wind.degree.degree.toDouble(),
                            WeatherContract.WeatherColumns.WindSpeedUnit.KPH
                        )
                    }
                    setForecast(forecastList)
                }
                mRequest!!.complete(ServiceRequestResult.Builder(builder.build()).build())
            }
        } catch (ignore: Exception) {
            requestWeatherFailed(requestLocation, RequestErrorType.WEATHER_REQ_FAILED)
        }
    }

    override fun requestWeatherFailed(requestLocation: Location, requestErrorType: RequestErrorType) {
        mRequest?.fail()
    }
}

internal object WeatherConditionConvertHelper {
    fun getConditionCode(code: WeatherCode?, dayTime: Boolean): Int {
        return when (code) {
            WeatherCode.CLEAR -> return if (dayTime) {
                WeatherContract.WeatherColumns.WeatherCode.SUNNY
            } else WeatherContract.WeatherColumns.WeatherCode.CLEAR_NIGHT
            WeatherCode.PARTLY_CLOUDY -> return if (dayTime) {
                WeatherContract.WeatherColumns.WeatherCode.PARTLY_CLOUDY_DAY
            } else WeatherContract.WeatherColumns.WeatherCode.PARTLY_CLOUDY_NIGHT
            WeatherCode.CLOUDY -> return WeatherContract.WeatherColumns.WeatherCode.CLOUDY
            WeatherCode.RAIN -> return WeatherContract.WeatherColumns.WeatherCode.SHOWERS
            WeatherCode.SNOW -> return WeatherContract.WeatherColumns.WeatherCode.SNOW
            WeatherCode.WIND -> return WeatherContract.WeatherColumns.WeatherCode.WINDY
            WeatherCode.FOG -> return WeatherContract.WeatherColumns.WeatherCode.FOGGY
            WeatherCode.HAZE -> return WeatherContract.WeatherColumns.WeatherCode.HAZE
            WeatherCode.SLEET -> return WeatherContract.WeatherColumns.WeatherCode.SLEET
            WeatherCode.HAIL -> return WeatherContract.WeatherColumns.WeatherCode.HAIL
            WeatherCode.THUNDER -> return WeatherContract.WeatherColumns.WeatherCode.THUNDERSTORMS
            WeatherCode.THUNDERSTORM -> return WeatherContract.WeatherColumns.WeatherCode.THUNDERSHOWER
            else -> WeatherContract.WeatherColumns.WeatherCode.NOT_AVAILABLE
        }
    }
}
